### ISSUE TITLE
Insecure example in ERC721 documentation

### ISSUE THREAD

**Issue by Alexander-Herranz** [OpenZeppelin/openzeppelin-contracts#4033]
Hi,

I noticed an insecure practice in the following example (_GameItem.sol_), from the documentation: https://docs.openzeppelin.com/contracts/4.x/erc721#constructing_an_erc721_token_contract

```
// contracts/GameItem.sol
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

import "@openzeppelin/contracts/token/ERC721/extensions/ERC721URIStorage.sol";
import "@openzeppelin/contracts/utils/Counters.sol";

contract GameItem is ERC721URIStorage {
    using Counters for Counters.Counter;
    Counters.Counter private _tokenIds;

    constructor() ERC721("GameItem", "ITM") {}

    function awardItem(address player, string memory tokenURI)
        public
        returns (uint256)
    {
        uint256 newItemId = _tokenIds.current();
        _mint(player, newItemId);
        _setTokenURI(newItemId, tokenURI);

        _tokenIds.increment();
        return newItemId;
    }
}

```

**💻 Environment**

Any

**📝 Details**

In the provided example you use the **_mint** function, that is said to be discouraged in the documentation:
https://docs.openzeppelin.com/contracts/4.x/api/token/erc721#ERC721-_mint-address-uint256-

</div><div class="paragraph" style="box-sizing: border-box; margin-left: 1em; margin-right: 1em; color: rgb(40, 40, 70); font-family: -apple-system, &quot;system-ui&quot;, &quot;avenir next&quot;, avenir, &quot;helvetica neue&quot;, helvetica, ubuntu, roboto, noto, &quot;segoe ui&quot;, arial, sans-serif; font-size: medium; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(254, 254, 254); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><br class="Apple-interchange-newline">Usage of this method is discouraged, use [_safeMint](https://docs.openzeppelin.com/contracts/4.x/api/token/erc721#ERC721-_safeMint-address-uint256-bytes-) whenever possible

**🔢 Code to reproduce bug**

You can [find the code in the documentation](https://docs.openzeppelin.com/contracts/4.x/erc721#constructing_an_erc721_token_contract).

**Comment by frangio** [comment]
I think I agree we should use `_safeMint`, but the `.increment()` line should be executed before minting.

**Comment by Rassska** [comment]
It seems that the function `awardItem(address, string memory)` is not supposed to be called by player, hence there is a risk of unbounded gas consumption. It also opens some doors to re-entrancy. So, if you're planning to replace it, be sure to include all details.

**Comment by Alexander-Herranz** [comment]
I think we can also replace the Counter with a regular uint++/uint--

Counters library was designed to avoid overflows when managing numbers, but I think that we do not need it anymore from Solidity 8.

**Comment by Medlab-web3** [comment]
You are correct in noticing that the _mint function is discouraged in the OpenZeppelin documentation in favor of using _safeMint. Here's why:

##  Why Use _safeMint Instead of _mint?
### Reentrancy Protection: 
The _safeMint function includes checks to ensure that the minting process is secure, especially when transferring tokens to addresses that are smart contracts. It prevents reentrancy attacks by ensuring that any contract receiving the minted token implements the onERC721Received function properly.

### Compatibility with Other Contracts: 
The _safeMint function ensures compatibility with contracts that may implement onERC721Received. This is important for ensuring that your token can be safely received by other contracts without causing unexpected behavior.

## Suggested Changes
Here's how you can modify your GameItem contract to use _safeMint instead of _mint:
```solidity
// contracts/GameItem.sol
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

import "@openzeppelin/contracts/token/ERC721/extensions/ERC721URIStorage.sol";
import "@openzeppelin/contracts/utils/Counters.sol";

contract GameItem is ERC721URIStorage {
    using Counters for Counters.Counter;
    Counters.Counter private _tokenIds;

    constructor() ERC721("GameItem", "ITM") {}

    function awardItem(address player, string memory tokenURI)
        public
        returns (uint256)
    {
        uint256 newItemId = _tokenIds.current();
        _safeMint(player, newItemId); // Use _safeMint instead of _mint
        _setTokenURI(newItemId, tokenURI);

        _tokenIds.increment();
        return newItemId;
    }
}

``` 
## Benefits of Using _safeMint
Safety: By using _safeMint, you automatically incorporate additional checks that protect against potential vulnerabilities, such as reentrancy attacks.
Better Developer Practices: Encourages Solidity developers to adopt best security practices from the beginning.
Improved Interoperability: _safeMint ensures that your contract is compatible with others that may require onERC721Received for proper handling of tokens.
*****Using _safeMint is recommended over _mint because it provides extra security, ensures compatibility with other contracts, and helps prevent potential vulnerabilities.

**Comment by TUPM96** [comment]
Opened https://github.com/OpenZeppelin/openzeppelin-contracts/pull/6549 to switch the ERC721 GameItem docs example to _safeMint while keeping the token id increment before minting.
